wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,327 @@
+	function saveInstance()
+		wantSave += 1
+
+	/** Releases Damage instance so it can be destroyed. */
+	function releaseInstance()
+		wantSave -= 1
+		if fired and wantSave <= 0
+			destroy this
+
+	protected int wantSave = 0
+	
+	/** stores how much health were added */
+	protected real afterDamage = 0.
+	protected bool killingBlow = false
+	private bool fired = false
+
+	/** applied all modifiers to the damage instance once */
+	function modify()
+		for i in damageModifiers
+			i.run(this)
+	protected function anounce()
+		for i in damageListeners
+			i.run(this)
+
+	/** Returns true if the damage is going to kill or already have killed target */
+	function isKillingBlow() returns bool
+		return killingBlow or (not fired and target.getHP() - .405 < amount)
+		
+	/** Multiplies initial damage with given magnitude */
+	function addMultiplier(real mult)
+		amount += initial * (mult - 1)
+
+	/** Increases damage amount with given amount */
+	function addAmount(real amount)
+		this.amount += amount
+		
+	construct(unit source, unit target, real amount, bool isSpell, int id)
+		this.source = source
+		this.target = target
+		this.initial = amount
+		this.amount = amount
+		spell = isSpell
+		this.id = id
+
+	protected static function processEvent() returns Damage
+		let damage = GetEventDamage()
+		if damage == 0
+			return null
+		let d = new Damage(GetEventDamageSource(), GetTriggerUnit(), (damage < 0 ? -damage : damage), damage < 0, DAMAGE_ID_NONTRIGGERED)
+		d.fired = true
+		return d
+
+	/** applies user-created Damage instance. */
+	function deal()
+		if not fired
+			fired = true
+			if target.isAlive()
+				let life = target.getHP()
+				killingBlow = life - .405 < amount
+				if killingBlow and doSaveTarget
+					amount = target.getHP() - .406
+					killingBlow = false			
+				dealCodeDamage(source, target, amount, killingBlow)
+				anounce()
+			if this.wantSave == 0
+				destroy this
+
+/** Deal an exact amount of damage to `target`, crediting kill experience and
+	gold bounty to `attacker` and not triggering any of DamageDetection listeners. KillingBlow parameter decides whether to kill the target. Isn't supposed to use outside of the system. */
+function dealCodeDamage(unit attacker, unit target, real damage, bool killingBlow)
+	if damage != 0
+		let hp = target.getHP()
+		disableDamageDetect()
+		if not killingBlow
+			target.setHP(max(hp - damage, .405))
+		else
+			// Deal damage using two separate components. The ATTACK_TYPE_UNIVERSAL
+			// case will deal damage to units of all armor types, even if that armor
+			// would otherwise block all damage. The ATTACK_TYPE_MAGIC case affect
+			// ethereal units, whereas other attack types do not.
+			target.setHP(.405)
+			attacker.damageTarget(target, 1000000., ATTACK_TYPE_UNIVERSAL)
+			attacker.damageTarget(target, 1000000., ATTACK_TYPE_MAGIC)
+		enableDamageDetect()
+
+function handler() returns bool
+	let d = Damage.processEvent()
+	if d == null
```

Comment:
I don't think this check ever does anything

Reply:
.processEvent() returns null if GetEventDamage == 0